src/permutation_test.js

```diff
@@ -0,0 +1,85 @@
+
+import mean from './mean';
+import shuffle from './shuffle';
+import sum from './sum';
+
+/**
+ * Conducts a [permutation test](https://en.wikipedia.org/wiki/Resampling_(statistics)#Permutation_tests)
+ * to determine if two data sets are *significantly* different from each other, using
+ * the difference of means between the groups as the test statistic. 
+ * The function allows for the following hypotheses:
+ * - two_tail = Null hypothesis: the two distributions are equal.
+ * - greater = Null hypothesis: observations from sampleX tend to be smaller than those from sampleY.
+ * - less = Null hypothesis: observations from sampleX tend to be greater than those from sampleY.
+ * [Learn more about one-tail vs two-tail tests.](https://en.wikipedia.org/wiki/One-_and_two-tailed_tests)
+ *
+ * @param {Array<number>} sampleX first dataset (e.g. treatment data)
+ * @param {Array<number>} sampleY second dataset (e.g. control data)
+ * @param {string} alternative alternative hypothesis, either 'two_sided' (default), 'greater', or 'less'
+ * @param {number} k number of values in permutation distribution. 
+ * @returns {number} p-value The probability of observing the difference between groups (as or more extreme than what we did), assuming the null hypothesis.
+ *
+ * @example
+ * var control = [2, 5, 3, 6, 7, 2, 5];
+ * var treatment = [20, 5, 13, 12, 7, 2, 2];
+ * permutationTest(control, treatment); // ~0.1324 
+ */
+function permutationTest(
+    sampleX/*: Array<number> */,
+    sampleY/*: Array<number> */,
+    alternative/*: string */ = 'two_side',
+    k/*: number */ = 10000)/*: ?number */ {
+    if (alternative !== 'two_side' && alternative !== 'greater' && alternative !== 'less') {
+        throw new Error('`alternative` must be either \'two_tail\', \'greater\', or \'less\'');
+    }
+
+    // get means for each sample
+    var meanX = mean(sampleX);
+    var meanY = mean(sampleY);
+    // int pValue var
+    var pValue;
+    
+    // calculate initial test statistic. This will be our point of comparison with
+    // the generated test statistics.
+    var testStatistic = meanX - meanY;
+    
+    // create test-statistic distribution
+    var testStatDsn = new Array(k);
+    
+    for (var i = 0; i < k; i++) {
+        
+        // 1. shuffle data assignments
+        var allData = sampleX.concat(sampleY);
+        var permData = shuffle(allData);
+        var permLeft = permData.splice(0, Math.floor(permData.length / 2));
+        var permRight = permData;
+    
+        // 2.re-calculate test statistic
+        var permTestStatistic = mean(permLeft) - mean(permRight);
+
+        // 3. store test statistic to build test statistic distribution
+        testStatDsn[i] = permTestStatistic;
+    }
+
+    // Calculate p-value depending on alternative 
+    // more info on p-value calculations: https://onlinecourses.science.psu.edu/stat464/node/35
+    if (alternative === 'two_side') {
+        pValue = sum(testStatDsn.map(function(tStat) {
```

Comment:
Here it would be nice not to have to create a new array to avoid memory allocation but this means rolling your own sum and it could get tedious.

Reply:
Since here all the parts of the sum are either `0` or `1`, we don't have to use the fancy error-correcting algorithm. It's both much faster and simpler to just use straight `for (...)` loops instead of `sum`/`map`.